### PR TITLE
On macOS, fix `WindowEvent::Destroyed` isn't emitted 

### PR DESCRIPTION
**Pull request by wusyong**
It seems `windowWillClose` can only be called when `windowShouldClose` returns true.
So the `WindowEvent::Destroyed` is never called when the window is dropped.
This PR add another queue event when the window is dropped during Drop trait is called.

- [x] Tested on all platforms changed
- [x] Added an entry to `CHANGELOG.md` if knowledge of this change could be valuable to users
- [x] Updated documentation to reflect any user-facing changes, including notes of platform-specific behavior
- [ ] Created or updated an example program if it would help users understand this functionality
- [ ] Updated [feature matrix](https://github.com/rust-windowing/winit/blob/master/FEATURES.md), if new features were added or implemented

### CODE REVIEW

File: src/platform_impl/macos/util/async.rs [hunk 1]
```diff
@@ -218,11 +222,17 @@ pub(crate) fn set_title_async(window: &NSWindow, title: String) {
...
     let window = MainThreadSafe(window);
     Queue::main().exec_async(move || {
         autoreleasepool(move |_| {
             window.close();
+
+            let event = Event::WindowEvent {
+                window_id: WindowId(window.id()),
+                event: WindowEvent::Destroyed,
+            };
+            AppState::queue_event(EventWrapper::StaticEvent(event));
```
Comment (madsmtm) [inline src/platform_impl/macos/util/async.rs:235]: Nit: Use `window.emit_event`

**Review by madsmtm — CHANGES_REQUESTED** [review]
Huh, odd that this hasn't been discovered before! Does this mean that our overriding of `windowWillClose:` is effectively useless?

However, I doubt this fixes the problem completely, since the window may be closed before the last reference to it is destroyed (for example in the `window_debug` example). Do you have an idea for how to fix this?

File: src/platform_impl/macos/util/async.rs (hunk 1 above)
Comment (wusyong) [reply src/platform_impl/macos/util/async.rs:235]: Only `WinitWindowDelegate` has this method. Should I add such method to `WinitWindow` too? Or could I just access `WinitWindowDelegate` by `pub(crate)` it.

**Comment by wusyong** [comment]
> Does this mean that our overriding of windowWillClose: is effectively useless?

I couldn't find any other way to trigger `windowWillClose:` than returning true in `windowShouldClose:`.

> However, I doubt this fixes the problem completely, since the window may be closed before the last reference to it is destroyed (for example in the window_debug example). Do you have an idea for how to fix this?

Yeah, `Destroyed` event won't be sent in this case but it's consistent across platforms (at least on x11 and windows). But I assume that's okay if you are already going to exit.  In my use case, I treat `Destroyed` event as a way to gracefully shutdown. I have a collection of window (HasMap for example). `CloseRequested` will drop the window in the collection and `set_exit` if the collection is empty during `Destroyed`.

File: src/platform_impl/macos/util/async.rs (hunk 1 above)
Comment (madsmtm) [reply src/platform_impl/macos/util/async.rs:235]: Just do `pub(crate)`

**Comment by madsmtm** [comment]
> I couldn't find any other way to trigger `windowWillClose:` than returning true in `windowShouldClose:`.

Then we should maybe remove the code in it, along with a note? Though I'm a bit weary about doing so, since it seems like, see this commit from 2018 that added the note about El Capitan https://github.com/goddessfreya/winit/commit/8e9fc01bd6b404f59488b130413f48e4e5f89b0d (from PR https://github.com/rust-windowing/winit/pull/853).

@francesca64 do you remember anything about this?

> Yeah, `Destroyed` event won't be sent in this case but it's consistent across platforms (at least on x11 and windows).

Would you mind documenting this on `Event::Destroyed`, then?

**Comment by francesca64** [comment]
@madsmtm I sadly don't remember, sorry!

**Comment by madsmtm** [comment]
No worries, thanks for the quick response!

File: src/platform_impl/macos/util/async.rs (hunk 1 above)
Comment (wusyong) [reply src/platform_impl/macos/util/async.rs]: @madsmtm I got a real weird behavior if I move a shared pointer of `delegate` into `exec_async`, `windowWillClose` will be called somehow!
So I can just call `delegate.window.close()` instead in latest commit. Wdyt?

Comment (wusyong) [reply src/platform_impl/macos/util/async.rs]: I found the delegate has to be used in `exec_async` somehow, so `windowWillCose` can be invoked.
So I call `delegate.window.close()` instead which should look a little bit cleaner.

**Comment by madsmtm** [comment]
I've discovered that `windowWillClose:` is called if you run `close:` synchronously, which means this should be fixed by https://github.com/rust-windowing/winit/pull/2574

**Comment by wusyong** [comment]
@madsmtm I slightly modified multiwindow example and check `windows` hash map on `Destroyed` event.
It can work based on your PR! I can close this one if you are going to merge #2574.

**Pull request closed without merging**